src/components/AuthPanel.tsx

```diff
@@ -38,21 +50,110 @@ export const AuthPanel = () => {
+            setErrorMessage(errorMessages.WRONG_EMAIL);
+            return;
+        }
+        setErrorMessage("");
+    }
+
+    // Check if postal code is valid format during onChange
+    const validatePostalCode = (input:string) => {
+        const canadianPostalCodeRegex =  /^[ABCEGHJ-NPRSTVXY]\d[ABCEGHJ-NPRSTV-Z][ -]?\d[ABCEGHJ-NPRSTV-Z]\d$/;
+        if (!canadianPostalCodeRegex.test(input))
+        {
+            setErrorMessage(errorMessages.INVALID_POSTAL_CODE);
+            return;
+        }
+        setErrorMessage("");
+    }
+
+    // Check if password and confirm password are the same
+    const validateConfirmPassword = (input:string) => {
+        if(password != input)
+        {
+            setErrorMessage(errorMessages.MISMATCH_PWD);
+            return;
+        }
+        //only check if password is valid during user registration
+        validatePassword(input);
+    }
+
+    const validatePassword = (input:string) => {
+        const passwordRegex = /^(?=.*[a-zA-Z])(?=.*[0-9]).{8,}$/;
+        if (!passwordRegex.test(input))
+        {
+            setErrorMessage(errorMessages.WRONG_PASSWORD);
+            return;
+        }
+        setErrorMessage("");
+    }
+    // Store Password on state for checking (during registration)
+    const saveAndValidatePassword = (input:string) => {
+        setPassword(input);
+        if(isLogin)
+        {
+            return;
+        }
+        //only check if password is valid during user registration
+        validatePassword(input);
+    }
+
     const handleSubmit = async (e: FormEvent<HTMLFormElement>) => {
         e.preventDefault();
         const formData = new FormData(e.currentTarget);
+        const inputEmail = formData.get("email")!.toString();
+        const inputPassword = formData.get("password")!.toString();
+        
+        if(inputEmail == "" || inputPassword == "")
+        {
+            setErrorMessage(errorMessages.REQUIRED);
+            return;
+        }
+        if (errorMessage != "")
+        {
+            return;
+        }
         const payload = {
-            email: formData.get("email")!.toString(),
-            password: formData.get("password")!.toString(),
+            email: inputEmail,
+            password: inputPassword
         }
         const registerPayload = {
             ...payload,
             postalCode: formData.get("postal_code")?.toString(),
         }
+        if(!isLogin)
+        {
+            const confirmPwd = formData.get("confirm_password")?.toString();
+            if(registerPayload.postalCode == "" || confirmPwd == undefined || confirmPwd == "")
+            {
+                setErrorMessage(errorMessages.REQUIRED);
+                return;
+            }
+            validatePassword(inputPassword);
+        }
+
+        if(errorMessage != "")
+        {
+            return;
+        }
         const res = isLogin ? await loginUser(payload) : await registerUser(registerPayload);
+        if(res == null)
```

Comment:
Can you also help me wrap the api call with a try catch block?
if it fails, it never gets to this line
so instead, we could set the error as follows:
```
try {
...
} catch (e){
console.log(e)
 setErrorMessage("Login Failed");
}
```

Reply:
Changed login failure from result checking to try-catch